### DISCUSSION TITLE
If DateTime is the primary key, can it search with time complexity O(logn)?

### DISCUSSION THREAD

**Discussion by Everesttour** [simolus3/drift#2439] [Q&A] [answered]
If DateTime is the primary key, can it search with time complexity O(logn)?
Is it also aligned?
I am a beginner. Thank you.

```
class Schedules extends Table {
  DateTimeColumn get startTime => dateTime()();
  IntColumn get category => integer()();
  TextColumn get content => text()();
  BoolColumn get isStored => boolean()();

  @override
  Set<Column> get primaryKey => {startTime};
}
```

**Comment by simolus3** [comment, marked as answer]
sqlite3 doesn't have a dedicated date column, so drift has two modes to store date times. As unix timestamps (the default), and as text ([which can have some advantages](https://drift.simonbinder.eu/docs/getting-started/advanced_dart_tables/#datetime-options)). The expected performance depends on the option you've chosen.

When storing values as an timestamps, there is a single-column integer primary key which becomes an alias to the rowid in sqlite3. Both exact lookups (`WHERE startTime = x`) and range queries (`WHERE startTime BETWEEN ...`) are heavily optimized by sqlite3 in that case. You can assume that this would be `O(log(N))` in both cases.
When storing datetimes as text, the index automatically created for primary keys will enable efficient (logarithmic) lookups for exact filters. Range filters can't be optimized in this mode though, an `BETWEEN` filter would have a linear complexity.

**Reply by Everesttour** [reply]
Thank you so much!!!
